subprojects/scheduler/src/main/java/org/gradle/internal/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,587 @@
+        if (numberOfParallelExecutors < 1) {
+            throw new IllegalArgumentException("Not a valid number of parallel executors: " + numberOfParallelExecutors);
+        }
+
+        this.executorCount = numberOfParallelExecutors;
+        this.workers = Lists.newArrayListWithCapacity(numberOfParallelExecutors);
+        this.availableWorkers = Queues.newArrayBlockingQueue(numberOfParallelExecutors);
+    }
+
+    @Override
+    public GraphExecutionResult execute(Graph graph, Collection<? extends Node> entryNodes, boolean continueOnFailure, NodeExecutor nodeExecutor) {
+        // TODO Get name via Gradle.findIdentityPath()
+        ManagedExecutor executor = executorFactory.create("Task worker for '" + "gradle" + "'");
+        try {
+            WorkerLease parentWorkerLease = workerLeaseService.getCurrentWorkerLease();
+            for (int index = 0; index < executorCount; index++) {
+                TaskExecutorWorker worker = new TaskExecutorWorker(index, coordinationService, parentWorkerLease, availableWorkers, eventQueue);
+                workers.add(worker);
+                executor.execute(worker);
+            }
+
+            ImmutableList.Builder<Node> executedNodes = ImmutableList.builder();
+            ImmutableList.Builder<Throwable> failures = ImmutableList.builder();
+
+            execute(graph, nodeExecutor, continueOnFailure, executedNodes, failures);
+            if (cancelled) {
+                failures.add(new BuildCancelledException());
+            }
+            return new GraphExecutionResult(executedNodes.build(), failures.build());
+        } finally {
+            for (TaskExecutorWorker worker : workers) {
+                worker.interrupt();
+            }
+            executor.stop();
+        }
+    }
+
+    private void cancelExecution(Graph graph, boolean cancelMustRun) {
+        for (Node node : graph.getAllNodes()) {
+            // Allow currently executing and enforced tasks to complete, but skip everything else.
+            switch (node.getState()) {
+                case RUNNABLE:
+                case SHOULD_RUN:
+                    node.setState(CANCELLED);
+                    cancelled = true;
+                    break;
+                case MUST_RUN:
+                    if (cancelMustRun) {
+                        node.setState(CANCELLED);
+                        cancelled = true;
+                    }
+                    break;
+                case CANCELLED:
+                case DEPENDENCY_FAILED:
+                    break;
+                default:
+                    throw new AssertionError();
+            }
+        }
+    }
+
+    private void execute(final Graph graph, final NodeExecutor nodeExecutor, final boolean continueOnFailure, ImmutableList.Builder<Node> executedNodes, ImmutableList.Builder<Throwable> failures) {
+        boolean expectAvailableWorkers = true;
+        while (graph.hasNodes()) {
+            if (expectAvailableWorkers) {
```

Comment:
What is the meaning of `expectAvailableWorkers`? That a worker should be available?
Then I don't understand why the `NodeSuspended` event returns `false` in the `handle` method. Shouldn't a worker be available if a node has been suspended?

Reply:
`NodeSuspendedEvent` was named unfortunately, and was used to signal a node that couldn't be scheduled yet. This has been changed now.